provider/equinix/environ.go

```diff
@@ -0,0 +1,649 @@
+func (e *environ) supportedInstanceTypes() ([]instances.InstanceType, error) {
+	opt := &packngo.ListOptions{
+		Includes: []string{"available_in"},
+	}
+	plans, _, err := e.equnixClient.Plans.List(opt)
+	if err != nil {
+		return nil, errors.Annotate(err, "retrieving supported instance types")
+	}
+
+	var instTypes []instances.InstanceType
+nextPlan:
+	for _, plan := range plans {
+		if !validPlan(plan, e.cloud.Region) {
+			continue
+		}
+
+		var instArch string
+		switch {
+		case strings.HasSuffix(plan.Name, ".x86"):
+			instArch = arch.AMD64
+		case strings.HasSuffix(plan.Name, ".arm"):
+			instArch = arch.ARM64
+		default:
+			continue nextPlan
+
+		}
+
+		mem, err := parseMemValue(plan.Specs.Memory.Total)
+		if err != nil {
+			continue
+		}
+
+		instTypes = append(instTypes,
+			instances.InstanceType{
+				Id:       plan.ID,
+				Name:     plan.Name,
+				CpuCores: uint64(plan.Specs.Cpus[0].Count),
+				Mem:      mem,
+				Arches:   []string{instArch},
+				// Scale per hour costs so they can be represented as an integer for sorting purposes.
+				Cost: uint64(plan.Pricing.Hour * 1000.0),
+				// TODO: returned by packet's API but not exposed by the packngo client
+				// Deprecated: plan.Legacy,
+			})
+	}
+
+	return instTypes, nil
+}
+
+func validPlan(plan packngo.Plan, region string) bool {
+	notAvailable := true
+	for _, a := range plan.AvailableIn {
+		if a.Code == region {
+			notAvailable = false
+			break
+		}
+	}
+	isInvalid := notAvailable || plan.Pricing == nil ||
+		plan.Specs == nil ||
+		plan.Specs.Memory == nil ||
+		len(plan.Specs.Cpus) == 0 || plan.Specs.Cpus[0].Count == 0
+
+	return !isInvalid
+}
+
+func parseMemValue(v string) (uint64, error) {
+	var scaler = uint64(1)
+	if strings.HasSuffix(v, "GB") {
+		scaler = 1024
+		v = strings.TrimSuffix(v, "GB")
+	}
+
+	val, err := strconv.ParseUint(v, 10, 64)
+	return val * scaler, err
+}
+
+func (e *environ) findInstanceSpec(controller bool, allImages []*imagemetadata.ImageMetadata, instanceTypes []instances.InstanceType, ic *instances.InstanceConstraint) (*instances.InstanceSpec, error) {
+	// version, err := series.UbuntuSeriesVersion(ic.Series)
+	// if err != nil {
+	// 	return nil, errors.Trace(err)
+	// }
+
+	oss, _, err := e.equnixClient.OperatingSystems.List()
+	if err != nil {
+		return nil, err
+	}
+	suitableImages := []*imagemetadata.ImageMetadata{}
+
+	for _, it := range instanceTypes {
+	nextImage:
+		for _, os := range oss {
+
+			switch os.Distro {
+			case "ubuntu":
+				series, err := series.VersionSeries(os.Version)
+				if err != nil || ic.Series != series {
+					continue nextImage
+				}
+			case "centos":
+				series, err := series.CentOSVersionSeries(os.Version)
+				if err != nil || ic.Series != series {
+					continue nextImage
+				}
+			case "windows":
+				series, err := series.WindowsVersionSeries(os.Version)
+				if err != nil || ic.Series != series {
+					continue nextImage
+				}
+			default:
+				continue nextImage
```

Comment:
Extract switch into method/function so it's more testable.

Reply:
The function only contains initializations and this logic. Is it really necessary to extract this into another function

Reply:
Yes, because it will be more testable. Large functions/methods means your coverage for this is going to be low.